keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1603 @@
+that does not guarantee that iptables will remain supported by
+_distributions_ forever. In particular, Red Hat has declared that
+[iptables is deprecated in RHEL 9] and is likely to be removed
+entirely in RHEL 10, a few years from now. Other distributions have
+made smaller steps in the same direction; for instance, [Debian
+removed `iptables` from the set of "required" packages] in Debian 11
+(Bullseye).
+
+The RHEL deprecation in particular impacts Kubernetes in two ways:
+
+  1. Many Kubernetes users run RHEL or one of its downstreams, so in a
+     few years when RHEL 10 is released, they will be unable to use
+     kube-proxy in `iptables` mode (or, for that matter, in `ipvs` or
+     `userspace` mode, since those modes also make heavy use of the
+     iptables API).
+
+  2. Several upstream iptables bugs and performance problems that
+     affect Kubernetes have been fixed by Red Hat developers over the
+     past several years. With Red Hat no longer making any effort to
+     maintain iptables, it is less likely that upstream iptables bugs
+     that affect Kubernetes in the future would be fixed promptly, if
+     at all.
+
+[iptables is deprecated in RHEL 9]: https://access.redhat.com/solutions/6739041
+[Debian removed `iptables` from the set of "required" packages]: https://salsa.debian.org/pkg-netfilter-team/pkg-iptables/-/commit/c59797aab9
+
+### The `ipvs` mode of kube-proxy will not save us
+
+Because of the problems with iptables, some developers added an `ipvs`
+mode to kube-proxy in 2017. It was generally hoped that this could
+eventually solve all of the problems with the `iptables` mode and
+become its replacement, but this never really happened. It's not
+entirely clear why... [kubeadm #817], "Track when we can enable the
+ipvs mode for the kube-proxy by default" is perhaps a good snapshot of
+the initial excitement followed by growing disillusionment with the
+`ipvs` mode:
+
+  - "a few issues ... re: the version of iptables/ipset shipped in the
+    kube-proxy container image"
+  - "clearly not ready for defaulting"
+  - "complications ... with IPVS kernel modules missing or disabled on
+    user nodes"
+  - "we are still lacking tests"
+  - "still does not completely align with what [we] support in
+    iptables mode"
+  - "iptables works and people are familiar with it"
+  - "[not sure that it was ever intended for IPVS to be the default]"
+
+Additionally, the kernel IPVS APIs alone do not provide enough
+functionality to fully implement Kubernetes services, and so the
+`ipvs` backend also makes heavy use of the iptables API. Thus, if we
+are worried about iptables deprecation, then in order to switch to
+using `ipvs` as the default mode, we would have to port the iptables
+parts of it to use nftables anyway. But at that point, there would be
+little excuse for using IPVS for the core load-balancing part,
+particularly given that IPVS, like iptables, is no longer an
```

Comment:
*question*
ahhh  I didnt know ipvs wasnt actively developed anymore.  cc @Damans227 .....

....  should we deprecate ipvs along the way? if we just whittle kube proxy down to one, NFT impl, that would make it super easy to refactor in the future? (maybe its a separate KEP, but , if were going all in on NFT, might as well deprecate iptables and IPVS together right) ?

Reply:
> Maybe linking to some link that represents the IPVS decline would be useful

Yeah, the problem is that the decline is more just of the "everyone stops paying attention to it" sort rather than the "formally deprecated" sort. There's just not a lot going on with that code, as compared to nftables (or eBPF). If you look at https://github.com/torvalds/linux/commits/master/net/netfilter/ipvs, there's a lot of "update the ipvs code to reflect this API change made elsewhere" type stuff, and a little bit of bugfixing (eg [this two-line fix](https://github.com/torvalds/linux/commit/d7fce52fdf96663ddc2eb21afecff3775588612a) that merged two weeks ago), but not really any new feature development.

> should we deprecate ipvs along the way?

That's an open question. As noted in the Motivation section, _this_ KEP is just about adding nftables (which will certainly result in us having to support all three of iptables, ipvs, and nftables for some amount of time). We'll have to continue to think about what comes next.

> IMo we could delete the windows proxy along the way ... if theres sufficient buy in from the community

That would actually help out some with the refactoring stuff that's going on in parallel with this KEP. Eg, as noted in https://github.com/kubernetes/kubernetes/pull/118464, there are simplifications we could make to the `pkg/proxy` code if we only had to support iptables/ipvs, but the winkernel backend does things differently enough that it makes things harder to refactor (especially since I can't even test the winkernel proxy easily).

OTOH, we already keep screwing up on "generic" vs "Linux-only" (https://github.com/kubernetes/kubernetes/issues/117909), and moving the Windows code out of tree would make that worse.

Maybe we should put both the windows proxy _and_ the nftables proxy out-of-tree...

Reply:
> Maybe we should put both the windows proxy and the nftables proxy out-of-tree...
Is that still open for debate? that would be nice and sorta help force the issue of making the kube proxy easier for people to understand, bc you'd have a reference, standalone impl w/o less  generic boiler plate.

Reply:
Since most of existing kube-proxy problems config versions, flags, Windows... require breaking compatibility (yeah, alpha means alpha, but this was alpha for years and will cause a lot of disruption across a lot of projects). Being able to start from scratch is worth considering

Reply:
though there are at least 2 other previous nftables implementations that I know of , and they went nowhere ... I like Dan's approach in the KEP,  the main driver is that `iptables` is going to be removed and become unmaintained

Reply:
he also explains this in L263, we can not become default this new mode if we break compatibility, users deployments will stop working